udata/api_fields.py

```diff
@@ -0,0 +1,247 @@
+        # the `List` RestX type with this converted inner value.
+        field_read, field_write = convert_db_to_field(f"{key}.inner", field.field, info.get('inner_field_info', {}))
+        constructor_read = lambda **kwargs: restx_fields.List(field_read, **kwargs)
+        constructor_write = lambda **kwargs: restx_fields.List(field_write, **kwargs)
+    elif isinstance(field, mongo_fields.ReferenceField):
+        # For reference we accept while writing a String representing the ID of the referenced model.
+        # For reading, if the user supplied a `nested_fields` (RestX model), we use it to convert
+        # the referenced model, if not we return a String (and RestX will call the `str()` of the model
+        # when returning from an endpoint)
+        nested_fields = info.get('nested_fields')
+        if nested_fields is None:
+            # If there is no `nested_fields` convert the object to the string representation.
+            constructor_read = restx_fields.String
+        else:
+            constructor_read = lambda **kwargs: restx_fields.Nested(nested_fields, **kwargs)
+
+        write_params['description'] = "ID of the reference"
+        constructor_write = restx_fields.String
+    elif isinstance(field, mongo_fields.EmbeddedDocumentField):
+        nested_fields = info.get('nested_fields')
+        if nested_fields is None:
+            raise ValueError(f"EmbeddedDocumentField `{key}` requires a `nested_fields` param to serialize/deserialize.")
+
+        constructor = lambda **kwargs: restx_fields.Nested(nested_fields, **kwargs)
+    else:
+        raise ValueError(f"Unsupported MongoEngine field type {field.__class__.__name__}")
+    
+    read_params = {**params, **read_params, **info}
+    write_params = {**params, **write_params, **info}
+
+    read = constructor_read(**read_params) if constructor_read else constructor(**read_params)
+    if write_params.get('readonly', False):
+        write = None
+    else:
+        write = constructor_write(**write_params) if constructor_write else constructor(**write_params)
+    return read, write
+
+def generate_fields(**kwargs):
+    '''
+    This decorator will create two auto-generated attributes on the class `__read_fields__` and `__write_fields__`
+    that can be used in API endpoint inside `except()` and `marshall_with()`.
+    '''
+    def wrapper(cls):
+        read_fields = {}
+        write_fields = {}
+        sortables = []
+
+        read_fields['id'] = restx_fields.String(required=True)
+
+        for key, field in cls._fields.items():
+            info = getattr(field, '__additional_field_info__', None)
+            if info is None: continue 
+
+            if info.get('sortable', False):
+                sortables.append(key)
+
+            read, write = convert_db_to_field(key, field)
+
+            if read:
+                read_fields[key] = read
+            if write:
+                write_fields[key] = write
+
+        for method_name in dir(cls):
```

Comment:
I think this entire section isn't quite clear on first read? Maybe move it to a dedicated method or add a comment?

Reply:
Is it better with 418eee698?

Reply:
Yep, thank you! :pray: